### PR TITLE
Implement RFC 2011 (nicer assert messages)

### PR DESCRIPTION
**Pull request by sinkuu**
#44838

```rust
assert!(x + y == z && w);
```

```
thread 'main' panicked at 'assertion failed: x + y == z && w
with expansion: 3 + 5 == 7 && (unevaluated)', test.rs:6:5
```

#### Detailed designs
:bike: :house: What should be captured? (Current implementation captures operands of unary and binary operators.) How should the panic message look like?

#### Performance
Now `assert` expands to some amount of code, and this may slow down compilation and execution. We need to test performance difference and see if it have to fall back to the original behavior when `not(debug_assertions)`.

### CODE REVIEW

**Comment by oli-obk** [comment]
> Current implementation captures operands of unary and binary operators.

I think function call arguments could be captured, too. There's no difference between `Add::add(&a, &b)` and `a + b` in my book.

**Comment by pietroalbini** [comment]
Highfive failed to assign a reviewer, assigning one randomly from the libs team.

r? @dtolnay

**Comment by sinkuu** [comment]
> I think function call arguments could be captured, too. There's no difference between Add::add(&a, &b) and a + b in my book.

I thought so too, but realized that they are different after trying to implement.

```rust
let x = &mut Foo;
let y = &mut Foo;

let _ = bar(x); // x is reborrowed
println!("{:?}", x);

let _ = x + y; // x is consumed
// println!("{:?}", x); // "use of moved value: `x`"
```

This is problematic because my strategy to capture an intermediate value by-value is to replace a expression with

```
{ let tmp = {expression}; try_copy(&tmp, &mut __capture); tmp }
```

which always consumes the expression.
I tried applying another transformation like `{ try_copy(&{path}, &mut __capture); func({path}) }` for `ExprKind::Path` (where reborrow works?) to preserve reborrow. This seemed to work at first, but bumped into `func(None)` (`None` is a `Path` on AST) causing type inference error. :expressionless:

**Comment by shepmaster** [comment]
Ping from triage, @dtolnay ! Will you have time to review this in the near future?

File: src/libcore/lib.rs [hunk 1]
```diff
@@ -208,3 +208,84 @@ pub use coresimd::simd;
...
+        #[unstable(feature = "generic_assert_internals", issue = "44838")]
+        pub fn new(value: Captured<T>, alt: &'static str) -> Self {
+            DebugFallback { value, alt }
+        }
+    }
+
+    #[unstable(feature = "generic_assert_internals", issue = "44838")]
+    impl<T> Debug for DebugFallback<T> {
+        default fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+            f.write_str(self.alt)
```
Comment (dtolnay) [inline src/libcore/lib.rs]: Is it possible for self.value to be Unevaluated here? If so, I would expect this to print `"(unevaluated)"` even for non-Debug types.

File: src/libsyntax_ext/assert.rs [hunk 2]
```diff
@@ -86,7 +761,7 @@ fn escape_format_string(s: &str) -> String {
 
 #[test]
 fn test_escape_format_string() {
-    assert!(escape_format_string(r"foo{}\") == r"foo{{}}\\");
+    assert!(escape_format_string("foo{}") == "foo{{}}");
```
Comment (dtolnay) [inline src/libsyntax_ext/assert.rs]: Is this change necessary? The implementation of `escape_format_string` has not changed.

File: src/libsyntax_ext/assert.rs [hunk 3]
```diff
@@ -35,40 +50,700 @@ pub fn expand_assert<'cx>(
...
+    }
+}
+
+#[derive(Clone)]
+struct CondExpr {
+    node: CondExprKind,
+    /// True if `||` or `&&` are contained in this tree.
+    contains_lazy: bool,
+    /// True if by-value captures are contained in this tree.
+    contains_by_ref: bool,
```
Comment (dtolnay) [inline src/libsyntax_ext/assert.rs]: The field name contradicts the doc comment (by-ref vs by-value).

**Review by dtolnay — COMMENTED** [review]
I still need to look at the implementation of Context::generate but here are some comments regarding the rest.

**Comment by dtolnay** [comment]
To help review, would you be able to share what src/libsyntax_ext/assert_tests.rs looks like after macro expansion? `rustc -Z unstable-options --pretty=expanded`

File: src/libcore/lib.rs (hunk 1 above)
Comment (sinkuu) [reply src/libcore/lib.rs]: Right, this needs to be

```rust
match self.value {
    Value | NotCopy => alt,
    Unevaluated => "(unevaluated)",
}
```

**Comment by sinkuu** [comment]
https://gist.github.com/sinkuu/ba6abd95cb55e839ae33cb022410bb36
(I added dummy `macro_rules! panic { ($($t:tt)*) => {panic($($t)*)}; }` to prevent explosion with `format!` expansion)

**Comment by ishitatsuyuki** [comment]
@dtolnay I guess that your review comments are addressed. Mind doing another review?

File: src/libcore/lib.rs [hunk 4]
```diff
@@ -208,3 +208,87 @@ pub use coresimd::simd;
 #[unstable(feature = "stdsimd", issue = "48556")]
 #[cfg(not(stage0))]
 pub use coresimd::arch;
+
+// FIXME: move to appropriate location
```
Comment (dtolnay) [inline src/libcore/lib.rs]: Please move this to libcore/asserting.rs (by analogy with runtime support for panicking in libcore/panicking.rs).

**Comment by dtolnay** [comment]
Thanks! I would be comfortable accepting this if for now we make it expand exactly the old way in release mode. Later we can debate what changes it makes sense to apply in release mode.

**Comment by ishitatsuyuki** [comment]
Is this insta-stable? I have no strong opinion but maybe it's good to have a feature flag to switch between new and old behaviour.

Regarding the performance concern, does it make sense to have a perf run (while we have this enabled as default)?

**Comment by sinkuu** [comment]
https://github.com/rust-lang/rust/pull/49071#issuecomment-375917182 applies to this PR, too:( I believe "plain text fallback" described in RFC2011 cannot be implmented without this specialization (`T` -> `T: Debug`).

**Comment by ishitatsuyuki** [comment]
This RFC can be backward compatible only if the broader specialization is used; do you think we can use that here? Is there a soundness issue? cc @nikomatsakis

Back to the implementation itself, can we track the evaluated status by using a counter? Short circuits has a defined ordering, and presumably computing that at codegen time allows better optimization.

**Comment by sinkuu** [comment]
I'm not sure whether if I get what you mean. Maybe listing all evaluation states and have branches of `panic` call for each of them? I'd avoid that as it would make explosion of the amount of the generated code much more worse. I think optimization for the case of assertion falure is not important and should not regress other factors.

**Comment by ishitatsuyuki** [comment]
Sounds like you're right. Never mind then.

**Comment by nikomatsakis** [comment]
@ishitatsuyuki 

> Is there a soundness issue? ... Sounds like you're right. Never mind then.

I'm not sure if my input is still needed, can you clarify? =)

**Comment by ishitatsuyuki** [comment]
@nikomatsakis As we are using a variant of specialization that isn't going to be stabilized, we need your input on whether it's acceptable to use such specialization or not. Also see #49071 which tries to achieve a similar specialization.

**Comment by sinkuu** [comment]
I will reopen once the discussion about sound specialization settles.
